scib/metrics/morans_i.py

```diff
@@ -0,0 +1,149 @@
+        # This should not happen, just in case
+        if (morans_i > 1).any() or (morans_i < -1).any():
+            raise ValueError(
+                "Problems in computing Moran's I, value not between -1 and 1"
+            )
+        data.var["morans_i"] = morans_i
+
+    if compare_pre:
+        # Get per sample connectivities
+        adatas_sample = dict()
+        for sample in adata_pre.obs[batch_key].unique():
+            # Subset only to cells from 1 sample
+            adata_sample = adata_pre[adata_pre.obs[batch_key] == sample, :].copy()
+            # Compute HVG for each sample
+            sc.pp.highly_variable_genes(
+                adata_sample,
+                flavor="cell_ranger",
+                batch_key="study_sample",
+                n_top_genes=2000,
+            )
+            adata_sample_scl = adata_sample.copy()
+            # PP each sample: scaling, pca, neighbours
+            sc.pp.scale(adata_sample_scl, max_value=10)
+            sc.pp.pca(
+                adata_sample_scl,
+                n_comps=15,
+                use_highly_variable=True,
+                svd_solver="arpack",
+            )
+            sc.pp.neighbors(adata_sample_scl, n_pcs=15)
+            # Add computed info back to unscaled data
+            adata_sample.uns["neighbors"] = adata_sample_scl.uns["neighbors"]
+            adata_sample.obsp["connectivities"] = adata_sample_scl.obsp[
+                "connectivities"
+            ]
+            adata_sample.obsp["distances"] = adata_sample_scl.obsp["distances"]
+            adata_sample.uns["pca"] = adata_sample_scl.uns["pca"]
+            adata_sample.obsm["X_pca"] = adata_sample_scl.obsm["X_pca"]
+            # Save adata of sample
+            adatas_sample[sample] = adata_sample
+        del adata_sample
+        del adata_sample_scl
+
+        # Compute Moran's I across samples and on integrated data
+        for data in list(adatas_sample.values()) + [adata_post]:
+            compute_mi(data, hvgs)
+
+        # Compute differences in Moran's I
+        # Table of Moran's I-s across batches
+        batch_mis = []
+        for sample, data in adatas_sample.items():
+            mi = data.var["morans_i"]
+            mi.name = sample
+            batch_mis.append(mi)
+        batch_mis = pd.concat(batch_mis, axis=1)
+        # Difference with integrated data
+        mi_diffs = adata_post.var["morans_i"] - batch_mis.max(axis=1)
+        avg_mi_diff = mi_diffs.mean()
+
+        # Rescale so that it is between [0,1] where 1 is better
+        # Moran's I will be in [-1,1] and thus difference can be in [-2,2]
+        if rescale:
+            res = (avg_mi_diff + 2) / 4
+        else:
+            res = avg_mi_diff
+
+    else:
+        # Moran's I on integrated data
+        compute_mi(adata_post, hvgs)
+        res = adata_post.var["morans_i"].mean()
+        if rescale:
+            # Moran's I will be in [-1,1]
+            res = (res + 1) / 2
```

Comment:
also here: should a Moran's I of 0 be better than one of -1? Could that difference only be due to sparsity of the marker?

Reply:
I changed it to scores >0 as we are mostly interested in clusterings.